visearch-android/src/main/java/com/visenze/visearch/android/network/SearchService.java

```diff
@@ -0,0 +1,186 @@
+            call = apiService.search(headers, params);
+        } else if(RECOMMENDATION.equals(endPoint)) {
+            call = apiService.recommendation(headers, params);
+        } else if (COLOR_SEARCH.equals(endPoint)) {
+            call = apiService.colorSearch(headers, params);
+        } else {
+            throw new ViSearchException("wrong API method");
+        }
+
+        call.enqueue(new Callback<ResponseData>() {
+            @Override
+            public void onResponse(Call<ResponseData> call, Response<ResponseData> response) {
+
+                if(response.isSuccessful() && response.body() !=null) {
+                    ResponseData data = response.body();
+                    handleResponse(data, resultListener);
+                } else {
+                    resultListener.onSearchError("api failed");
+                }
+            }
+
+            @Override
+            public void onFailure(Call<ResponseData> call, Throwable t) {
+                resultListener.onSearchError(t.getMessage());
+            }
+        });
+    }
+
+
+    public void uploadSearch(String endPoint, UploadSearchParams uploadSearchParams, final ViSearch.ResultListener resultListener) {
+        byte[] imageBytes = null;
+
+        if (uploadSearchParams.getImage() != null) {
+            imageBytes = uploadSearchParams.getImage().getByteArray();
+        }
+
+        String imageUrl = uploadSearchParams.getImageUrl();
+        String imId = uploadSearchParams.getImId();
+
+        if (imageBytes == null && (imageUrl == null || imageUrl.isEmpty()) && (imId == null || imId.isEmpty())) {
+            throw new ViSearchException("Missing parameter, image empty");
+        }
+
+        Map<String, String> headers = getHeaders();
+        Call<ResponseData> call;
+        if(imageBytes != null) {
+            RequestBody imageBody = RequestBody.create(MediaType.parse("image/*"), imageBytes);
+            MultipartBody.Part image = MultipartBody.Part.createFormData("image", "image", imageBody);
+            Map<String, RequestBody> params = buildPostParams(uploadSearchParams.toMap());
+            if(UPLOAD_SEARCH.equals(endPoint)) {
+                call = apiService.uploadSearch(headers, image, params);
+            } else if (DISCOVER_SEARCH.equals(endPoint)) {
+                call = apiService.discoverSearch(headers, image, params);
+            } else {
+                throw new ViSearchException("wrong API method");
+            }
+
+        } else {
+            Map<String, String> params = buildGetParams(uploadSearchParams.toMap());
+            if(UPLOAD_SEARCH.equals(endPoint)) {
+                call = apiService.uploadSearch(headers, params);
+            } else if(DISCOVER_SEARCH.equals(endPoint)) {
+                call = apiService.discoverSearch(headers, params);
+            } else {
+                throw new ViSearchException("wrong API method");
+            }
+
+        }
+
+        call.enqueue(new Callback<ResponseData>() {
```

Comment:
this seem to be exactly the same as line 57, please refactor into 1 common method

Reply:
changed already